src/nvim/normal.c

```diff
@@ -4897,6 +4897,10 @@ static void nv_next(cmdarg_T *cap)
     (void)normal_search(cap, 0, NULL, SEARCH_MARK | cap->arg, NULL);
     cap->count1 -= 1;
   }
+  // Redraw old and new line to ensure that the highlighting of the current match
+  // under the cursor as well as the old match both get properly updated.
+  redrawWinline(curwin, old.lnum);
+  redrawWinline(curwin, curwin->w_cursor.lnum);
```

Comment:
This can redraw the wrong line if the line number of the old match changed. Not sure how to deal with this.

Reply:
Clearing the entire window is the only way I can think of

Reply:
@bfredl thoughts?

Reply:
I used `redrawWinline` on `curwin->w_last_cursorline` because bfredl suggested me, not fully sure if it fixes the issue though. Having a way to test it would be nice but I can't think of any

Reply:
It seems `curwin->w_last_cursorline` only gets updated when `cursorline_standout` is true so it wouldn't be very reliable to redraw that line, so I just removed the redraw for that line. I honestly don't see a case where the old match can change in the middle of using N or n, so I think we'd be fine even without a way to mitigate that. @zeertzjq can you think of a case where that might happen?

Reply:
When one pastes lines above cursor.

Reply:
> It can change
> when one pastes lines above cursor, but that not in the middle of using n/N.

That means we don't need to consider that possibility since the value of old line is taken only after using n/N, so old match would have to change while `nv_next` is running, which seems impossible to me

Reply:
But search highlight doesn't disappear unless one stops it manually.

Reply:
> But search highlight doesn't disappear unless one stops it manually.

But nv_next doesn't deal with that, does it?

Reply:
It seems the highlighting works perfectly fine even with all of the redraw statements in `nv_next` removed, it seems they weren't really necessary in the first place since we already handle things in `update_search_hl`

Reply:
I completely changed how highlights are updated because I noticed a bug where CurSearch wasn't being updated when the cursor was moved out of the match, so now we don't need to redraw in `nv_next` at all, which resolves this problem